### ISSUE TITLE
`ignored-unit-patterns` ignores mutability

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#11754]
### Summary

.

### Lint Name

ignored-unit-patterns

### Reproducer

I tried this code:

```rust
// edition:2021
// check-pass
#![feature(coroutines)]

fn main() {
    let x = &mut ();
    || {
        let _c = || yield *&mut *x;
        || _ = &mut *x;
    };
}

```

I saw this happen:

```
warning: matching over `()` is more explicit
 --> issue-110929-coroutine-conflict-error-ice.rs:9:12
  |
9 |         || _ = &mut *x;
  |            ^ help: use `()` instead of `_`: `()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ignored_unit_patterns
  = note: requested on the command line with `-W clippy::ignored-unit-patterns`
```

Applying the suggestion breaks the code.

### Version

```text
rustc 1.75.0-nightly (9c20ddd95 2023-11-03)
binary: rustc
commit-hash: 9c20ddd956426d577d77cb3f57a7db2227a3c6e9
commit-date: 2023-11-03
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.4
```
